
line0=${PBS_ARRAYID}          # Line number in the bamfiles list
OUTPUT=/path/to/output        # Output directory for storing intermediate BAMs and final phased BAM
REF=susScr11.fa               # Reference genome 
INPUT=/path/to/phased_vcf     # Directory containing per-chromosome SHAPEIT5 phased VCFsï¼ŒGenerated by the previous step code 03.shapeit5.sh
THREADS=20                    # Number of threads for samtools, bcftools, and WhatsHap

mkdir -p ${OUTPUT}
cd ${OUTPUT}

ID=$(sed -n "${line0}p" ${OUTPUT}/F2.JR.bamfiles | awk '{print $1}')
BAM=$(sed -n "${line0}p" ${OUTPUT}/F2.JR.bamfiles | awk '{print $2}')
ID2=$(echo $ID | sed 's/.$//' | fgrep -wf - ${OUTPUT}/F2.infor | awk '{print "F2-"$2}')

chroms=( $(awk '{print $1}' ${ref}.fai | sed -n 1,20p | tr '\n' ' ') )
echo "[$(date)] Chromosomes: ${chroms[*]}"

for CHR in "${chroms[@]}"; do
{
    echo "[$(date)] Processing ${CHR}"

    # Extract sample from phased VCF
    bcftools view -s ${ID2} ${INPUT}/${CHR}.shapiet.vcf.gz -Oz --threads=${THREADS} \
        -o ${ID2}.${CHR}.vcf.gz
    tabix -p vcf ${ID2}.${CHR}.vcf.gz

    # Add read group
    RG="@RG\\tID:${ID2}\\tPL:RNAseq\\tSM:${ID2}\\tLB:${ID2}\\tPU:1"

    # Extract per-chr BAM
    samtools view -h ${BAM} -@ ${THREADS} ${CHR} \
        | samtools addreplacerg -r ${RG} - -@${THREADS} \
        | samtools view -Sb -@${THREADS} \
        > ${ID}.${CHR}.bam
    samtools index ${ID}.${CHR}.bam -@${THREADS}

    # Haplotag with WhatsHap
    singularity exec ~/01.Biosoft/03.Sif/whatshap.simg \
        /opt/conda/envs/whatshap-env/bin/whatshap haplotag \
        --regions ${CHR} \
        -o ${ID}.${CHR}.phased.bam \
        -r ${ref} \
        ${ID2}.${CHR}.vcf.gz \
        ${ID}.${CHR}.bam
    samtools index ${ID}.${CHR}.phased.bam -@${THREADS}

} &
done && wait

ls ${ID}.*.phased.bam > ${ID}.bamfiles2
samtools cat -b ${ID}.bamfiles2 -o ${ID2}.RNAphased.bam
samtools index -@${THREADS} ${ID2}.RNAphased.bam

##
ID=`sed -n "${line0}p" ${OUTPUT}/RNAphased.bamfiles|awk '{print $1}'`
BAM=`sed -n "${line0}p" ${OUTPUT}/RNAphased.bamfiles|awk '{print $2}'`

####
samtools stats $BAM -@20 >$ID.RNAphased.stats &

samtools  view -h $BAM  -@20 |awk '{if($0~/HP:i:1/ || $1~/@/)print }' | samtools view -h -Sb -@10 >$ID.Hap1.bam
samtools index $ID.Hap1.bam -@20
samtools stats $ID.Hap1.bam  -@20 >$ID.Hap1.stats

samtools  view -h $BAM  -@20 |awk '{if($0~/HP:i:2/ || $1~/@/)print }' | samtools view -h -Sb -@10 >$ID.Hap2.bam
samtools index $ID.Hap2.bam -@20
samtools stats $ID.Hap2.bam  -@20 >$ID.Hap2.stats

