#!/usr/bin/env bash
set -euo pipefail


# --------------------------
# Parameters
# --------------------------
OUTPUT=/path/to/output
INPUT=/path/to/phased_BAMs
GTF=Sus_scrofa.Sscrofa11.1.100.gtf
THREADS=40

# Select BAM list file
BAMLIST= F2.JR.Phased.bamfiles    # Generated by the previous step code 04.WhatsHap_haplotag.sh
###Example lines in the list:
path/F2-1000.RNAphased.bam
path/F2-1000.Hap1.bam
path/F2-1000.Hap2.bam

# --------------------------
# Loop over all samples in the list
# --------------------------
while read -r ID BAM; do

    singularity exec -B ${INPUT}:${INPUT} /home/guilu/software/star_2.7.9a_subread.simg featureCounts \
        -T ${THREADS} \
        -p -t gene \
        -g gene_id \
        -a ${GTF} \
        -o ${OUTPUT}/${ID}.featureCounts.txt \
        -s 2 \
        ${INPUT}/${BAM}

    echo "[$(date)] Finished sample: ${ID}"
done < ${BAMLIST}

# Generate gene list (Geneid + Length)
awk 'NR>2 {print $1"\t"$6}' \
    ${Inpath2}/F2.featureCounts.txt \
> gene.list

for i in F2-*.featureCounts.txt; do
    id=$(basename "$i" .featureCounts.txt)

    # Total (RNAphased / All reads)
    cut -f7 "$i" | sed '1,2d' > ${id}.countAll

    # Hap1
    cut -f7 ${Inpath2}/${id}.Hap1.featureCounts.txt | sed '1,2d' > ${id}.countHap1

    # Hap2
    cut -f7 ${Inpath2}/${id}.Hap2.featureCounts.txt | sed '1,2d' > ${id}.countHap2
done

paste gene.list F2-*.countAll > merged_All_counts.txt
paste gene.list F2-*.countHap1 > merged_Hap1_counts.txt
paste gene.list F2-*.countHap2 > merged_Hap2_counts.txt


