#!/usr/bin/env bash
set -euo pipefail


# --------------------------
# Parameters
# --------------------------
OUTPUT=/path/to/output
INPUT=/path/to/phased_BAMs
GTF=Sus_scrofa.Sscrofa11.1.100.gtf
THREADS=40

# Select BAM list file
BAMLIST= F2.JR.Phased.bamfiles    # Generated by the previous step code 04.WhatsHap_haplotag.sh
###Example lines in the list:
path/F2-1000.RNAphased.bam
path/F2-1000.Hap1.bam
path/F2-1000.Hap2.bam

# --------------------------
# Loop over all samples in the list
# --------------------------
while read -r ID BAM; do
    singularity exec -B ${INPUT}:${INPUT} /home/guilu/software/139.stringtie2.sif stringtie \
    -p 40 \
    -e \
    -B \
    -G $GTF \
    -o ${ID}.gtf \
    -A ${ID}.tsv \
    $BAM \
    --fr
    singularity exec -B ${INPUT}:${INPUT} /home/guilu/software/star_2.7.9a_subread.simg featureCounts \
        -T ${THREADS} \
        -p -t gene \
        -g gene_id \
        -a ${GTF} \
        -o ${OUTPUT}/${ID}.featureCounts.txt \
        -s 2 \
        ${INPUT}/${BAM}

    echo "[$(date)] Finished sample: ${ID}"
done < ${BAMLIST}

#TPM matrix
name1=F2
ls ./ |grep  "tsv$" >all.sample.geneCounts.files
for i in `cat all.sample.geneCounts.files `
do
        name=`echo $i |awk -F . '{print $1}'`
        csvtk uniq -f 1 -Tt $i >$i.rmdup
        cat $i.rmdup | grep -Ev "chrX|chrY|chrM|NW" |awk '{print $1,$NF}' |sed 's/gene://g' |sed "s/TPM/$name/g" |tr " " "\t" |csvtk uniq -f 1 -Tt  >${i%%.*}.
done
csvtk join -Tt `ls *TMP.tmp` >$name1.TPM.txt
#gene Counts matrix
ls ./ |grep  "featureCounts.txt$" >all.sample.featureCounts.files
for i in `cat all.sample.featureCounts.files`
do
         dirname `fgrep "Geneid" $i |awk '{print $NF}'`  |sed 's/\//\\\//g' |awk '{print $0"\\\/"}' >$i.dir.tmp
         dir=`cat $i.dir.tmp`
         bufix_raw=`fgrep "Geneid" $i |awk -F "/" '{print $NF}'`
         bufix=`fgrep "Geneid" $i |awk -F "/" '{print $NF}' |awk -F . '{print $1}'`     
         cat $i | grep -Ev "chrX|chrY|chrM|NW" | awk '{print $1,$NF}' | sed "s/$dir//g" |sed "s/$bufix_raw/$bufix/g" |sed 's/gene://g' |fgrep -v "#" |tr " " "

done
csvtk join -Tt `ls *FC.tmp` >$name1.GeneCounts.txt
#rm *tmp

